Reactive Messaging context service configuration

Azquelt opened the pull request:
Several changes here:

- Remove the RM 3.0 dependency on concurrent
  - Use global thread pool and internal `WSContextService` instead of  `ManagedExecutorService` and `ManagedScheduledExecutorService`
- Add explicit tests for context propagation with RM
- Add new config option to select a custom context service

With this change, users who are using the `concurrent-x.x` feature, can define a context service in server.xml:

```xml
    <contextService id="myContextService">
        <classloaderContext/>
        <jeeMetadataContext/>
    </contextService>
```

They can then configure the kafka connector to use it by setting MP Config properties. It can be set globally:
```properties
mp.messaging.connector.liberty-kafka.context.service=myContextService
```

It can also be set for a specific channel:
```properties
mp.messaging.incoming.myChannel.context.service=myContextService
```

If this property is not set, it will use the default context service if `concurrent-x.x` is enabled, or the built-in context service otherwise.

Fixes #26678

Azquelt committed:
Reactive Messaging context service configuration

Allow a custom context service to be configured for use by the Kafka
connector when running async tasks.

Don't rely on the default ManagedExecutorService.

Azquelt committed:
RM Context service test updates

Update existing Reactive Messaging tests for the context service
changes.

Azquelt committed:
Remove RM 3.0 dependency on concurrent

Since it no longer depends on ManagedExecutorService, we can remove the
dependency on the concurrent feature.

Azquelt committed:
Add Reactive Messaging context propagation tests

Add explicit tests that context is propagated correctly when using the
Kafka connector.

Add tests for the new option to specify a custom context service.

Azquelt committed:
RM no longer depends on concurrent: test updates

Add concurrent feature to tests which still need it.

abutch3r commented on dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/PartitionTracker.java:
```diff
@@ -72,9 +74,10 @@ public boolean isClosed() {
      * @return a CompletionStage which completes when any associated processing has been completed (e.g. when the message offset has been committed)
      */
-    public CompletionStage<Void> recordDone(long offset, Optional<Integer> leaderEpoch) {
+    public CompletionStage<Void> recordDone(long offset, Optional<Integer> leaderEpoch, RMContext context) {
```
How is the context here being applied to the result, given we currently aren't doing anything with it.

Azquelt replied:
Here the context is unused. We don't need to apply it in this method because we return a `CompletableFuture` which is already complete.

However, this method takes the context as an argument because the method is overridden in `CommittingPartitionTracker` which returns an incomplete `CompletableFuture` and applies the context around the call to `complete()` when that's called asynchronously.

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,183 @@
+     * The liberty global thread pool executor
+     */
+    @Reference(target = "(component.name=com.ibm.ws.threading)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+//    @Reference(service = ManagedExecutorService.class, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
```
Is this commented out item needed, looks like it was has been replaced?

Azquelt replied:
This is left over from when I was tracking down the cause of a behaviour change when using the global thread pool vs. `ManagedExecutorService`. I will remove it.

Azquelt commented:
> I can't find how the new tests are going to get run. there are no changes in any of the parent test classes in the `suite` package to either add a new Context set or an update to an existing one

Yeah, good spot, they should have been added to `PlaintextTests`. I was testing with the suite modified to only run the new tests, then I undid my changes to run all the tests again without actually committing the change to add the new test classes.

Azquelt commented:
#build
#libby

I think I've addressed all the comments, this is ready for re-review.

abutch3r reviewed (dismissed):
Assuming the Tests pass on the latest PB, LGTM

abutch3r commented:
All failures are unrelated to Reactive messaging and have confirmed that the new tests have been run in the rtc with all repeats

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+import io.openliberty.microprofile.reactive.messaging.internal.interfaces.RMContext;
+
+@Component
+public class RMAsyncProviderFactoryImpl implements RMAsyncProviderFactory {
```
Set `configurationPolicy` for this component

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+    @Reference(target = "(id=DefaultContextService)", cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile WSContextService defaultContextService;
+
+    private final Map<String, WSContextService> namedContextServices = Collections.synchronizedMap(new HashMap<>());
```
Add Javadoc, maybe use `ConcurrentHashMap`?

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+    protected void removeContextService(WSContextService contextService) {
+        namedContextServices.entrySet().removeIf(e -> e.getValue() == contextService);
+    }
```
This seems simpler:
```suggestion
    protected void removeContextService(WSContextService contextService) {
        namedContextServices.values().remove(contextService);
    }
```

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+        namedContextServices.entrySet().removeIf(e -> e.getValue() == contextService);
+    }
+
+    private String getContextServiceName(Map<String, Object> properties) {
+        Object name = properties.get("id");
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+        return new NamedAsyncProvider(contextServiceRef);
+    }
+
+    @SuppressWarnings("unchecked")
+    private RMContext captureContext(String contextServiceName) {
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    private class NamedAsyncProvider implements RMAsyncProvider {
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+        private final ThreadContextDescriptor context;
+
+        public RMContextImpl(ThreadContextDescriptor context) {
+            this.context = context;
+        }
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/interfaces/QuiesceParticipant.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ *
+ */
+public interface QuiesceParticipant {
```
Add Javavdoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/interfaces/QuiesceRegister.java:
```diff
@@ -0,0 +1,17 @@
+    void register(QuiesceParticipant participant);
+
+    void remove(QuiesceParticipant participant);
+}
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/interfaces/RMAsyncProviderFactory.java:
```diff
@@ -0,0 +1,18 @@
+public interface RMAsyncProviderFactory {
+
+    RMAsyncProvider getAsyncProvider(String contextServiceRef);
+}
```
Add Javadoc, including how to obtain an instance.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/QuiesceRegisterImpl.java:
```diff
@@ -0,0 +1,48 @@
+ *
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class QuiesceRegisterImpl implements ServerQuiesceListener, QuiesceRegister {
```
Add Javadoc

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/QuiesceRegisterImpl.java:
```diff
@@ -0,0 +1,48 @@
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class QuiesceRegisterImpl implements ServerQuiesceListener, QuiesceRegister {
+
+    private final Set<QuiesceParticipant> participants = new HashSet<>();
```
I think this should be a `ConcurrentHashMap` or a synchronized set to ensure thread safety.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/QuiesceRegisterImpl.java:
```diff
@@ -0,0 +1,48 @@
+    public void serverStopping() {
+        for (QuiesceParticipant participant : participants) {
+            participant.quiesce();
+        }
```
Consider thread safety here also.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/interfaces/RMAsyncProvider.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Provides the services needed for Reactive Messaging components to do asynchronous tasks
+ */
+public interface RMAsyncProvider {
```
Doc how to obtain an instance.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/interfaces/RMContext.java:
```diff
@@ -0,0 +1,28 @@
+ *******************************************************************************/
+package io.openliberty.microprofile.reactive.messaging.internal.interfaces;
+
+/**
+ * Represents a context which can be applied to a thread
+ */
+public interface RMContext {
```
Doc how to obtain an instance.

Azquelt reviewed (commented):
Noticed some missing doc and a possible thread-safety issue.

Azquelt commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+    @SuppressWarnings("unchecked")
+    private RMContext captureContext(String contextServiceName) {
+        WSContextService namedContextService = null;
+        if (contextServiceName != null) {
+            namedContextService = namedContextServices.get(contextServiceName);
+        }
+
```
If a context service is not found for a given name, (i.e. `namedContextServices.get` reutrns `null`) no error or warning is reported.

Azquelt committed:
Review updates to RM Async impl

- Improve Javadoc
- Address potential thread safety issues

njr-11 commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -0,0 +1,226 @@
+    /**
+     * Updates {@link #namedContextServices} when a new service is added
+     */
+    @Reference(service = WSContextService.class, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MULTIPLE)
+    protected void addContextService(WSContextService contextService, Map<String, Object> properties) {
+        String name = getContextServiceName(properties);
+        if (name != null) {
+            namedContextServices.put(name, contextService);
+        }
+    }
```
This seems inefficient to be tracking and receiving notifications for every configured contextService, including the ones you don't need and will never use.  Does this have any advantage over simply looking for the one you want in the service registry when you need it and tracking only that instance?
Also, if you keep it this way, it seems you are later manually filtering out the instances without an id. You could write a target filter that requires a non-null id, and then OSGi will only give you the ones with ids and you can remove the manual code.

Azquelt replied:
My thinking was:
- we need to find a context service by name regularly (usually at least once per message processed)
  - we need to look it up every time we use it because context services defined in the server.xml can change dynamically
- the total number of context services will be small
- context services don't come and go very often

I assumed (without testing) that it would be quicker to maintain a map from name to context service than to do an OSGi service lookup every time we wanted to use it. I wasn't aware of an easy way to do a lookup the first time we need a context service with a given name and then track it to see if it goes away.

I guess we could create a `ServiceTracker` every time we need a service with a given name and maintain a map of name to corresponding `ServiceTracker`? The tracker could remove itself from the map if the service goes away. I was assuming that the total number of context services would be small so the overhead of maintaining the map would be insignificant and didn't consider doing something more complex.

The suggestion to use a filter to ignore services without an ID is a reasonable one.

njr-11 replied:
I made the comment because it seemed odd to me to be collecting and tracking extra services that aren't needed, but your assumptions here are valid and so your approach does make sense.  Most other users of ContextService have server.xml config (or app config) that directly references a single instance or defaults, in which case the dependency can be enforced in OSGi on the one instance that is needed.  You have a more complex situation with the ability to receive requests for multiple different context services.

njr-11 reviewed (commented):
I added a few comments/questions.

Also, I should point out that server.xml isn't the only way for users to define a ContextService. As of concurrent-3.0, they can also define a ContextService within the application by using `@ContextServiceDefinition` or `<context-service>` in the deployment descriptor.

```
@ContextServiceDefinition(name = "java:module/env/MyContextService", ...
```

When a user does that, Liberty writes equivalent configuration via ConfigAdmin. The resulting configuration will have
  name = "java:module/env/MyContextService"
  application = "MyApp" // whichever app it came from
  module = "MyModule" // whichever module it came from
  id = "application[MyApp]/module[MyModule]/contextService[java:module/env/MyContextService]"
  ...
(You might want to double check the above before relying on it because I wrote the example from memory)
The main thing to point out here is that although it has a generated id, the user doesn't know that id. Instead, the user knows it by the name they gave it and which application, module, and component they defined it in.
The solution that you have looks like it will sort of work with this pattern (whether you intended it to or not) but has a gap where the user probably doesn't know which value to supply as context.service

Azquelt commented:
> Also, I should point out that server.xml isn't the only way for users to define a ContextService. As of concurrent-3.0, they can also define a ContextService within the application by using `@ContextServiceDefinition` or `<context-service>` in the deployment descriptor.
> 
> ```
> @ContextServiceDefinition(name = "java:module/env/MyContextService", ...
> ```
> 
> When a user does that, Liberty writes equivalent configuration via ConfigAdmin. The resulting configuration will have
> ```
> name = "java:module/env/MyContextService"
> application = "MyApp" // whichever app it came from
> module = "MyModule" // whichever module it came from
> id = "application[MyApp]/module[MyModule]/contextService[java:module/env/MyContextService]"
> ...
> ```
> (You might want to double check the above before relying on it because I wrote the example from memory)
>
> The main thing to point out here is that although it has a generated id, the user doesn't know that id. Instead, the user knows it by the name they gave it and which application, module, and component they defined it in. The solution that you have looks like it will sort of work with this pattern (whether you intended it to or not) but has a gap where the user probably doesn't know which value to supply as context.service

Yeah, I didn't consider that and wasn't assuming that the application-defined ContextServices would be available as OSGi services. Supporting application-defined context services wasn't included in the design, though I did have a think about it. I assumed I'd need to look up the `ContextService` from JNDI and have a second implementation of `RMAsyncProvider` which could capture context using a `ContextService`.

We're always looking up a context service within the context of an application, so we could extend the definition of the `context.service` parameter so that it can be used to select an application-defined `ContextService` from the current application.

Is that something we need to consider doing here?

njr-11 commented:
> Is that something we need to consider doing here?

The most important thing will be to ensure it doesn't partly work, after which we will be stuck with needing to preserve that behavior. It should either be completely disallowed or properly supported.

It looks like you very nearly have it supported already but would need to figure out what the correct externals of it are. We would either need to document the way of forming the id that can be used, or you would need to allow a JNDI name to be configured as the context.service and then either obtain the application/module/component if needed from the J2EEName on the thread, or if that isn't possible then provide a way for the user to supply the application/module/component.

njr-11 commented:
I don't want to hold up progress on this with my comment about application-defined ContextServices.  One option would be to temporarily force that code path to fail, allowing that work to be split out and done/decided separately.

Azquelt commented:
> The most important thing will be to ensure it doesn't partly work, after which we will be stuck with needing to preserve that behavior. It should either be completely disallowed or properly supported.

Yes, that's a good point.

> One option would be to temporarily force that code path to fail, allowing that work to be split out and done/decided separately.

I will do this for now by excluding services where the `application` and `module` properties are set. I'll raise another issue to decide whether to leave it like that or implement support for application-defined context services.

Azquelt committed:
RM: Explicitly ignore app-defined context services

The current design allows for context services to be specified in
server.xml and selected by name from MP Config.

However, application-defined Context Services (defined using
@ContextServiceDefinition from Jakarta Concurrency 3.0) are also
registered as OSGi services and the current code would allow those to
also be looked up by name, if the user knows the way liberty constructs
the names.

As this scenario is untested and unsupported, explicitly ignore any
application-defined context services.

Co-authored-by: Nathan Rauh <nathan.rauh@us.ibm.com>

njr-11 commented:
> will do this for now by excluding services where the `application` and `module` properties are set. I'll raise another issue to decide whether to leave it like that or implement support for application-defined context services.

For application-defined context services that are in the java:app scope, only the `application` property will be present, not `module`.

For application-defined context services that are in the java:global scope, neither `application` nor `module` properties will be present, although in this case, if someone puts in the java:global JNDI name as the context.service value, then I expect it will work.

njr-11 commented on dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java:
```diff
@@ -130,6 +130,16 @@ private String getContextServiceName(Map<String, Object> properties) {
+     * @return {@code true} if the service properties are for an application-defined context service, otherwise {@code false}
+     */
+    private boolean isApplicationDefinedContextService(Map<String, Object> properties) {
+        return properties.containsKey("application") && properties.containsKey("module");
```
Identify application-defined context services in java:global, java:app, and java:comp as well as those in java:module
```suggestion
        return !"file".equals(properties.get("config.source"));
```

Azquelt replied:
Thanks, I'll squash this in

Azquelt commented:
#libby
#build
